proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+
+For example, consider the following record struct:
+```C#
+record struct R1(T1 P1, T2 P2);
+```
+
+For this record struct, the synthesized equality members would be something like:
+```C#
+struct R1 : IEquatable<R1>
+{
+    public T1 P1 { get; init; }
+    public T1 P1 { get; init; }
+    public override bool Equals(object? obj) => obj is R1 && Equals((R1)obj);
+    public virtual bool Equals(R1 other)
+    {
+        return
+            EqualityComparer<T1>.Default.Equals(P1, other.P1) &&
+            EqualityComparer<T2>.Default.Equals(P2, other.P2);
+    }
+    public static bool operator==(R1 r1, R1 r2)
+        => r1.Equals(r2);
+    public static bool operator!=(R1 r1, R1 r2)
+        => !(r1 == r2);    
+    public override int GetHashCode()
+    {
+        return Combine(
+            EqualityComparer<T1>.Default.GetHashCode(P1),
+            EqualityComparer<T2>.Default.GetHashCode(P2));
+    }
+}
+```
+
+### Printing members: PrintMembers and ToString methods
+
+The record struct includes a synthesized method equivalent to a method declared as follows:
+```C#
+private bool PrintMembers(System.Text.StringBuilder builder);
+```
+
+The method does the following:
+1. for each of the record struct's printable members (non-static public field and readable property members), appends that member's name followed by " = " followed by the member's value: `this.member`, separated with ", ",
+2. return true if the record struct has printable members.
+
+The `PrintMembers` method can be declared explicitly.
+It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+The record struct includes a synthesized method equivalent to a method declared as follows:
+```C#
+public override string ToString();
+```
+
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+The synthesized method:
+1. creates a `StringBuilder` instance,
+2. appends the record struct name to the builder, followed by " { ",
+3. invokes the record struct's `PrintMembers` method giving it the builder, followed by " " if it returned true,
+4. appends "}",
+5. returns the builder's contents with `builder.ToString()`.
+
+For example, consider the following record struct:
+
+``` csharp
+record struct R1(T1 P1, T2 P2);
+```
+
+For this record struct, the synthesized printing members would be something like:
+
+```C#
+struct R1 : IEquatable<R1>
+{
+    public T1 P1 { get; init; }
+    public T2 P2 { get; init; }
+
+    protected virtual bool PrintMembers(StringBuilder builder)
+    {
+        builder.Append(nameof(P1));
+        builder.Append(" = ");
+        builder.Append(this.P1); // or builder.Append(this.P1); if P1 has a value type
```

Comment:
Am I missing something here, what is the difference when `P1` has a value type?

Reply:
Should be `P1.ToString()` when a value type.

Reply:
```suggestion
        builder.Append(this.P1); // or builder.Append(this.P1.ToString()); if P1 has a value type
```